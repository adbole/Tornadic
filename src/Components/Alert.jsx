import PropTypes from 'prop-types'
import { Widget } from "./BaseComponents";

const Alert = (props) => (
    <Widget className={"alert " + props.type}>
        <h2>{props.name}</h2>
        <p>{props.message}</p>
    </Widget>
)
Alert.propTypes = {
    type: PropTypes.oneOf(['warning', 'watch', 'advisory', 'special']).isRequired,
    name: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired,
    moreExist: PropTypes.bool
}

export default Alert;