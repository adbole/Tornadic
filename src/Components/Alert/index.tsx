import React from "react";
import booleanPointInPolygon from "@turf/boolean-point-in-polygon";

import { useBooleanState, usePermission } from "Hooks";

import { useWeather } from "Contexts/WeatherContext";

import AlertModal from "Components/Modals/Alert";

import type NWSAlert from "ts/NWSAlert";

import AlertWidget, { AlertInformation, ExcessAlerts } from "./style";


const getPriorityAlert = (alerts: NWSAlert[]) =>
    alerts.reduce(
        (highest, next) => (next.priority < highest.priority ? next : highest),
        alerts[0]
    );

function Alert({ noNotify = false }) {
    const { alerts: unfilteredAlerts, point } = useWeather();
    const notiPermission = usePermission("notifications")

    const [modalOpen, showModal, hideModal] = useBooleanState(false);

    const alerts = unfilteredAlerts.filter(alert => {
        //Point and NWSAlert include the necessary GeoJSON properties for this to work so conversion to any is okay
        const inBounds = alert.hasCoords() && booleanPointInPolygon(point as any, alert as any)
            
        return alert.get("affectedZones").includes(point.properties.forecastZone) || inBounds
    });

    const previousAlertTime = React.useRef(new Date(alerts[0]?.get("sent") ?? 0));

    React.useEffect(() => {
        if (noNotify || alerts.length === 0 || notiPermission !== "granted") return;

        const newAlerts = alerts.filter((alert) => new Date(alert.get("sent")) > previousAlertTime.current);

        if (newAlerts.length > 0) {
            const alert = getPriorityAlert(newAlerts);

            const additionalAlerts = newAlerts.length > 1 ? `\n+${newAlerts.length - 1} more alert(s)` : "";

            new Notification(`${alert.get("event")}`, {
                body: `Issued: ${alert.get("sent")}\nExpires: ${alert.get("expires") ?? alert.get("ends")}${additionalAlerts}`,
            });

            previousAlertTime.current = new Date(newAlerts[0].get("sent"));
        }

    //This effect should only run if alerts changes, noNotify and notiPermission are excluded as changing them could 
    //cause unnecessary notifications to be sent annoying the user.
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [alerts]);

    if (!alerts.length) return null;

    const alert = getPriorityAlert(alerts);

    return (
        <>
            <AlertWidget
                type={alert.getAlertCSS()}
                isTemplate
                size="widget-wide"
                onClick={showModal}
            >
                <AlertInformation>
                    <h1>{alert.get("event")}</h1>
                    <p>
                        <em>Issued:</em> {alert.get("sent")}
                    </p>
                    <p>
                        <em>Until:</em> {alert.get("ends") ?? alert.get("expires")}
                    </p>
                </AlertInformation>

                {alerts.length > 1 && (
                    <ExcessAlerts>+{alerts.length - 1} more alert(s)</ExcessAlerts>
                )}
            </AlertWidget>
            <AlertModal alerts={alerts} isOpen={modalOpen} onClose={hideModal} />
        </>
    );
}

export default Alert;
