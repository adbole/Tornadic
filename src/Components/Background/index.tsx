import React from "react";
import { Global } from "@emotion/react";
import styled from "@emotion/styled";
import { Sky } from "@react-three/drei";
import { Canvas } from "@react-three/fiber";

import { useReadLocalStorage } from "Hooks";

import { useWeather } from "Contexts/WeatherContext";

import { Clouds, Gradient, RainSnow, Stars, Thunder } from "./__internal__";


const StyledCanvas = styled(Canvas)({
    top: 0,
    left: 0,
    zIndex: -1,
});

function HighConstrast({ rayleigh }: { rayleigh: number }) {
    const settings = useReadLocalStorage("userSettings");

    if (rayleigh !== 1 || !settings?.highContrastForLive) return null;

    return <Global styles={{ "#root": { color: "#3f3f3f" } }} />;
}

export default function Background() {
    const { weather } = useWeather();
    const settings = useReadLocalStorage("userSettings");

    const { isDay, condition } = React.useMemo(
        () => ({
            isDay: weather.isDay(),
            condition: weather.getNow().conditionInfo.type,
        }),
        [weather]
    );

    const [rayleigh, elevation, exposure] = React.useMemo(() => {
        if (isDay) {
            switch (condition) {
                case "Clear":
                case "Mostly Clear":
                    return [1, 0.75, 0.25];
                default:
                    return [5, 0.75, 0.1];
            }
        }

        return [0.01, 0.75, 0.2];
    }, [isDay, condition]);

    return (
        <>
            <HighConstrast rayleigh={rayleigh} />
            <Gradient isDay={isDay} condition={condition} />
            
            {settings && !settings.preferGradient && (
                <StyledCanvas
                    style={{ position: "fixed" }}
                    gl={{ toneMappingExposure: exposure }}
                >
                    <Sky
                        rayleigh={rayleigh}
                        inclination={elevation}
                        mieCoefficient={0.005}
                        mieDirectionalG={0.99}
                        turbidity={1}
                        azimuth={0.25}
                    />
                    <Clouds isDay={isDay} condition={condition} />
                    <RainSnow condition={condition} />
                    <Thunder condition={condition} />
                    <Stars isDay={isDay} condition={condition} />
                </StyledCanvas>
            )}
        </>
    );
}
