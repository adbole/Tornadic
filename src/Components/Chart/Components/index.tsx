export { default as Area } from "./Area"
export { default as Axes } from "./Axes"
export { default as Bar } from "./Bar"
export { default as Line } from "./Line"
export { default as NowReference } from "./NowReference"
export { default as Tooltip, useTooltip } from "./Tooltip"