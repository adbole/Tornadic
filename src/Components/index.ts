export { default as Alert } from "./Alert";
export { default as Daily } from "./Daily";
export { default as HazardLevel } from "./HazardLevel";
export { default as Hourly } from "./Hourly";
export * from "./Input";
export { default as MessageScreen } from "./MessageScreen";
export { default as Chart } from "./Modals/Chart";
export { default as Modal } from "./Modals/Modal";
export * from "./Modals/Modal";
export { default as Now } from "./Now";
export { default as Pressure } from "./Pressure";
export { default as Radar } from "./Radar";
export { default as SimpleInfoWidget } from "./SimpleInfoWidget";
export { default as SkeletonWidget } from "./SkeletonWidget";
export { default as SunTime } from "./SunTime";
export { default as Widget } from "./Widget";
export { default as Wind } from "./Wind";
