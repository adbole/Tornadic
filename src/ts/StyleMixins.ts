import { css } from "@emotion/react";


export const mediaQueries = {
    large: "@media screen and (min-width: 1400px)",
    mediumMin: "@media screen and (min-width: 1100px)",
    mediumMax: "@media screen and (max-width: 1100px)",
    smallMin: "@media screen and (min-width: 800px)",
    smallMax: "@media screen and (max-width: 800px)",
} as const;

export const vars = {
    background: "var(--background)",
    backgroundLayer: "var(--background-layer)",
    borderRadius: "var(--border-radius)",
    inputBorderRadius: "var(--input-border-radius)",

    svgSize: "var(--svg-size)",

    primary: "var(--primary)",

    zLayer1: "var(--z-layer-1)",
    zLayer2: "var(--z-layer-2)",
} as const;

export const varNames= {
    background: "--background" as any,
    backgroundLayer: "--background-layer" as any,
    borderRadius: "--border-radius" as any,
    inputBorderRadius: "--input-border-radius" as any,

    svgSize: "--svg-size" as any,

    primary: "--primary" as any,

    zLayer1: "--z-layer-1" as any,
    zLayer2: "--z-layer-2" as any,
} as const;

export const alertColors = {
    warning: {
        background: "#C31700",
        foreground: "white",
    },
    watch: {
        background: "#FFF501",
        foreground: "black",
    },
    advisory: {
        background: "#FF9431",
        foreground: "black",
    },
    statement: {
        background: "#8F00FF",
        foreground: "white",
    },
    none: {
        background: "gray",
        foreground: "black",
    },
};


export const darkBackBlur = css({
    backgroundColor: "rgba(0, 0, 0, 0.75)",
    backdropFilter: "blur(10px)",
    color: "white",
});

export const centerFlex = css({
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
});

export const interactable = css({
    transition: "filter 0.3s ease",
    filter: "brightness(100%)",

    "&:hover, &:focus": { cursor: "pointer" },

    "&:active, &:disabled": { filter: "brightness(70%)" },
});

