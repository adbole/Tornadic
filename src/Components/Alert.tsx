import React from 'react';
import { AlertType, WeatherData } from '../ts/WeatherData';
import { NWSAlert, useWeather } from './Contexes/WeatherContext';
import { Modal, useModal } from './Contexes/ModalContext';
import { Widget, WidgetSize } from './SimpleComponents';

/**
 * Takes an ISO8601 string and converts it to a nice readable string in the format weekday, month, day, hour:minute AM/PM timezone
 * @param iso The ISO8601 string to convert
 * @returns The now nice readable string obtained from conversion
 */
function GetTimeString(iso: string) {
    return new Date(iso).toLocaleTimeString("en-us", {weekday:"short", month:"short", day:"numeric", hour12:true, hour:"numeric", minute:"numeric", timeZoneName:"short"});
}

function ToAlertCSS(alert: AlertType) {
    switch(alert) {
        case AlertType.WARNING:
            return "alert-warning";
        case AlertType.WATCH:
            return "alert-watch";
        case AlertType.ADVISORY:
            return "alert-advisory";
        default:
            return "alert-none";
    }
}

const AlertDisplay = ({alertData, className, onClick} : {alertData: NWSAlert, className: string, onClick: React.MouseEventHandler<HTMLDivElement>}) => (
    <Widget size={WidgetSize.WIDE} id="alert" className={className} onClick={onClick}>
        <h2>{alertData.properties.event}</h2>
        <p>{alertData.properties.event} until {GetTimeString(alertData.properties.ends ?? alertData.properties.expires)}</p>
    </Widget>
);

const Alert = () => {
    const alertData = useWeather().alerts;
    const modals = useModal();

    //If no alerts are active then don't display this component.
    if(!alertData.length) return <></>;

    //Determine which alert should be shown. Warnings get highest priority with the lastest being first. 
    let alertToShow = alertData[0];
    let highestAlertType = WeatherData.GetAlertType(alertData[0]);
    let newestAge = new Date(alertData[0].properties.sent);

    for(let i = 1; i < alertData.length; ++i) {
        const ithType = WeatherData.GetAlertType(alertData[i]);
        const sentDate = new Date(alertData[i].properties.sent);

        //Replace the alertToShow if its severity is lower
        if(highestAlertType < ithType) {
            alertToShow = alertData[i];
            highestAlertType = ithType;
            newestAge = sentDate;
        }
        //When two alerts are the same severity, we want the youngest one
        else if(highestAlertType === ithType && sentDate > newestAge) {
            alertToShow = alertData[i];
            highestAlertType = ithType;
            newestAge = sentDate;
        }
    }

    const onClickHandler = () => modals.showModal(alertData.length > 1 ? <AlertSelectionModal alertData={alertData}/> : <AlertModal alert={alertData[0]}/>);

    return <AlertDisplay alertData={alertToShow} className={ToAlertCSS(highestAlertType)} onClick={onClickHandler}/>;
};

export default Alert;

const AlertSelectionModal = ({alertData}: {alertData: NWSAlert[]}) => {
    const modals = useModal();

    //Memoize the component to prevent unnecessary operations
    const memoizedComponent = React.useMemo(() => (
        <Modal modalTitle={alertData.length + " Weather Alerts"} modalTitleClass={"alert-none"} id="alert-list-modal">
            {
                alertData.map((alert, key) => {
                    //Inclues override to onClose to reopen the alert selection modal
                    const onClickHandler = () => modals.showModal(<AlertModal alert={alert} onClose={() => modals.showModal(memoizedComponent)}/>);

                    return <AlertDisplay key={key} alertData={alert} className={ToAlertCSS(WeatherData.GetAlertType(alert))} onClick={onClickHandler}/>;
                })
            }
        </Modal>
    ), [alertData]);
    

    return memoizedComponent;
};

/**
 * Display the given NWSAlert as a modal to get more in depth information from.
 */
export const AlertModal = (props: {alert: NWSAlert} & React.DialogHTMLAttributes<HTMLDialogElement>) => {
    const {alert, ...excess} = props;

    const alertData = alert.properties;
    const alertType = ToAlertCSS(WeatherData.GetAlertType(alert));
    
    return (
        <Modal modalTitle={alertData.event} modalTitleClass={alertType} id="alert-modal" {...excess}>
            <p><em>Issuing Office:</em> {alertData.senderName}</p>
            <p><em>Issued:</em> {GetTimeString(alertData.sent)}</p>
            <p><em>Effective:</em> {GetTimeString(alertData.effective)}</p>
            <p><em>Ends:</em> {GetTimeString(alertData.ends ?? alertData.expires)}</p>
            <hr/>
            {
                alertData.instruction && 
                <>
                    <h3>Instructions</h3>
                    <p>{alertData.instruction}</p>
                    <hr/>
                </>
            }
            {
                alertData.description.split("*").map((string, index) => {
                    if(!string.length) return <React.Fragment key={index}/>;
        
                    return <p key={index}>* {string}</p>;
                })    
            }
            <hr/>
            <h3>Affected Areas</h3>
            <p>{alertData.areaDesc}</p>
        </Modal>
    );
};