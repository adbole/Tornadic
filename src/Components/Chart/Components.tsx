import { AreaChart, BarChart, LineChart, TooltipProps } from "recharts";
import { CategoricalChartProps } from "recharts/types/chart/generateCategoricalChart";
import { NameType } from "recharts/types/component/DefaultTooltipContent";

import { UV_MAX_VALUES } from "ts/Constants";

import { ChartViews, DataPoint } from ".";

function getUnit(data: DataPoint, defaultUnit: React.ReactNode) {
    const { property, primaryKey } = data;

    if(property === ChartViews.UV_Index) {
        if (primaryKey <= UV_MAX_VALUES.LOW) return " Low"; 
        else if (primaryKey <= UV_MAX_VALUES.MODERATE) return " Moderate";
        else if (primaryKey <= UV_MAX_VALUES.HIGH) return " High";
        else if (primaryKey <= UV_MAX_VALUES.VERY_HIGH) return " Very High";
        else return " Extreme"; 
    }

    return defaultUnit;
}
  
//Customized tooltip to display primary and secondary data with labels if applicable
export const CustomTooltip = ({ active, payload }: TooltipProps<number, NameType>) => {
    if(!active || !payload || !payload.length) return null;

    const fixDecimal = (property: ChartViews, value: number): string | number => 
        (property === ChartViews.Precipitation || property === ChartViews.Pressure) ? value.toFixed(2) : value.toFixed(0);

    const data: DataPoint = payload[0].payload;
    let unit = getUnit(data, payload[0].unit);
    const label = (data.property === ChartViews.Temperature && "Feels: ") || (data.property === ChartViews.Windspeed && "Gust: ");


    return (
        <div className="chart-tooltip">
            <h1>{fixDecimal(data.property, data.primaryKey)}{unit}</h1>
            {data.secondaryKey && <p>{label}{fixDecimal(data.property, data.secondaryKey)}{unit}</p>}
        </div>
    );
};

type ChartDisplayProps = {
    children: React.ReactNode,
    property: ChartViews
} & CategoricalChartProps

//Provides the proper chart to display based on the HourlyProperty passed
export const ChartDisplay = (props: ChartDisplayProps) => {
    const { children, property, ...excess } = props;

    let ChosenChart = BarChart;

    switch(property) {
        case ChartViews.Precipitation:
            ChosenChart = BarChart;
            break;
        case ChartViews.Temperature:
        case ChartViews.UV_Index:
            ChosenChart = AreaChart;
            break;
        default:
            ChosenChart = LineChart;
    }

    return (
        <ChosenChart {...excess}>
            {children}
        </ChosenChart>
    );
};