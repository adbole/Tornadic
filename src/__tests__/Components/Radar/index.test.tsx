import testIds from "__tests__/__constants__/testIDs"
import { render, screen } from "@testing-library/react"

import { Radar } from "Components"

import { mediaQueries } from "ts/StyleMixins"


const mocks = vi.hoisted(() => ({
    AttributionControl: ({ position, prefix }: { position: string, prefix: string }) => (
        <span data-position={position} data-prefix={prefix}>AttributionControl</span>
    ),
    MapContainer: vi.fn(({ children }: { children: React.ReactNode }) => (
        <div>
            <span>MapContainer</span>
            {children}
        </div>
    )),
    TileLayer: ({ position, prefix, className }: { position: string, prefix: string, className: string }) => (
        <span data-position={position} data-prefix={prefix} className={className}>TileLayer</span>
    ),
    ZoomControl: vi.fn(() => <span>ZoomControl</span>),
    ControlPortal: vi.fn(({ position, children }: { position: string, children: React.ReactNode }) => (
        <div data-position={position}>
            <span>ControlPortal</span>
            {children}
        </div>
    )),
    Opacity: vi.fn(() => <span>Opacity</span>),
}))

vi.mock("svgs/widget", async (importOriginal) => ({
    ...(await importOriginal() as any),
    Map: () => <span>Map</span>,
}))

vi.mock("react-leaflet", async (importOriginal) => ({
    ...(await importOriginal() as any),
    AttributionControl: mocks.AttributionControl,
    MapContainer: mocks.MapContainer,
    TileLayer: mocks.TileLayer,
    ZoomControl: mocks.ZoomControl,
}))

vi.mock("Components/Radar/__internal__", async (importOriginal) => ({
    ...(await importOriginal() as any),
    AlertPolygons: () => <span>AlertPolygons</span>,
    ControlPortal: mocks.ControlPortal,
    Home: () => <span>Home</span>,
    Locate: () => <span>Locate</span>,
    Opacity: mocks.Opacity,
    Peek: () => <span>Peek</span>,
    RainViewer: () => <span>RainViewer</span>,
}))

test("Renders as expected", () => {
    render(<Radar />)

    //Existance
    expect.soft(screen.queryByText("MapContainer")).toBeInTheDocument()
    expect.soft(screen.queryByText("TileLayer")).toBeInTheDocument()
    expect.soft(screen.queryByText("ZoomControl")).toBeInTheDocument()
    expect.soft(screen.queryByText("AttributionControl")).toBeInTheDocument()

    expect.soft(screen.queryByText("AlertPolygons")).toBeInTheDocument()
    expect.soft(screen.queryByText("ControlPortal")).toBeInTheDocument()
    expect.soft(screen.queryByText("Home")).toBeInTheDocument()
    expect.soft(screen.queryByText("Locate")).toBeInTheDocument()
    expect.soft(screen.queryByText("Opacity")).toBeInTheDocument()
    expect.soft(screen.queryByText("Peek")).toBeInTheDocument()
    expect.soft(screen.queryByText("RainViewer")).toBeInTheDocument()

    //Map props
    expect.soft(mocks.MapContainer).toHaveBeenLastCalledWith(
        expect.objectContaining({
            center: [35.5, -97.5],
            zoom: 10,
            zoomControl: false,
            attributionControl: false,
            scrollWheelZoom: false,
            dragging: false,
        }),
        expect.anything()
    )
})

test("Matches Snapshot", () => {
    const { container } = render(<Radar />)

    expect(container).toMatchSnapshot()
})

test("Radar has grid-area r down to medium screens", () => {
    render(<Radar />)

    expect(screen.getByTestId(testIds.Widget.WidgetSection))
        .toHaveStyleRule("grid-area", "r", { media: mediaQueries.min("medium") })
})