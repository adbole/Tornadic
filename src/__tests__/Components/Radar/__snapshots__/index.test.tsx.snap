// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Matches Snapshot 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  padding: 10px;
  overflow: hidden;
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  box-shadow: inset 0 0 0 200px rgba(255, 255, 255, 0.01);
  border: 1px solid rgba(255, 255, 255, 0.1);
  border-radius: var(--border-radius);
  min-height: 300px;
  grid-column: span 2;
  grid-row: span 2;
  -webkit-backdrop-filter: none;
  backdrop-filter: none;
}

.emotion-1>:not(.emotion-4) {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-1 .leaflet-center {
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-1 .leaflet-control {
  background-color: rgba(0, 0, 0, 0.75);
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  color: white;
  border-radius: var(--border-radius);
  overflow: hidden;
}

.emotion-1 .leaflet-custom-control,
.emotion-1 .leaflet-control-toggle {
  background-color: rgba(0, 0, 0, 0.75);
  -webkit-backdrop-filter: blur(10px);
  backdrop-filter: blur(10px);
  color: white;
  color: white!important;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 2rem;
  min-height: 2rem;
  border: none;
  --svg-size: 1.25rem;
}

.emotion-1 .leaflet-overlay-pane svg {
  width: unset;
}

@media screen and (min-width: 1100px) {
  .emotion-1 {
    grid-area: r;
  }
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 5px;
  font-size: 1rem;
  margin-bottom: 10px;
}

.emotion-5 {
  -webkit-filter: invert(1) hue-rotate(180deg);
  filter: invert(1) hue-rotate(180deg);
}

<div>
  <section
    class="emotion-0 emotion-1 emotion-2"
    data-testid="WIDGET_SECTION"
  >
    <h1
      class="emotion-3 emotion-4"
    >
      <span>
        Map
      </span>
       
      Radar
    </h1>
    <div>
      <span>
        MapContainer
      </span>
      <span
        class="emotion-5 emotion-6"
      >
        TileLayer
      </span>
      <span>
        ZoomControl
      </span>
      <span
        data-position="topleft"
        data-prefix="<a href="https://leafletjs.com/" title="A JavaScript library for interactive maps">Leaflet</a>"
      >
        AttributionControl
      </span>
      <span>
        AlertPolygons
      </span>
      <div
        data-position="leaflet-top leaflet-right"
      >
        <span>
          ControlPortal
        </span>
        <span>
          Home
        </span>
        <span>
          Locate
        </span>
        <span>
          Opacity
        </span>
      </div>
      <span>
        Peek
      </span>
      <span>
        RainViewer
      </span>
    </div>
  </section>
</div>
`;
